pkg/server/server.go

```diff
@@ -621,6 +628,26 @@ func (s *applicationServer) AuthorizeGitlab(ctx context.Context, msg *pb.Authori
 	return &pb.AuthorizeGitlabResponse{Token: token}, nil
 }
 
+func (s *applicationServer) ValidateProviderToken(ctx context.Context, msg *pb.ValidateProviderTokenRequest) (*pb.ValidateProviderTokenResponse, error) {
+	token, err := middleware.ExtractProviderToken(ctx)
+	if err != nil {
+		return nil, grpcStatus.Error(codes.Unauthenticated, err.Error())
+	}
+
+	v, err := findValidator(msg.Provider, s)
+	if err != nil {
+		return nil, grpcStatus.Error(codes.InvalidArgument, err.Error())
+	}
+
+	if err := v.ValidateToken(ctx, token.AccessToken); err != nil {
+		return nil, grpcStatus.Error(codes.InvalidArgument, err.Error())
+	}
+
+	return &pb.ValidateProviderTokenResponse{
+		Valid: true,
```

Comment:
Am I right in thinking we don’t currently differentiate between “we couldn’t check this token for some reason” and “this token is definitively not valid”? Would that be useful, or is that over-engineered for our use case?

Reply:
We have different error messages based on these cases:

"We couldn't find a token" -> `codes.Unauthenticated`

"We couldn't find a git provider for the name you specified" -> `codes.InvalidArg`

"The git provider says this token is no good" -> `codes.InvalidArg`

The first case is "we couldn't check this", the second two are "we checked it and its not valid". The code returned *could* be used to control the behavior on the client side.

I decided to let the error messages do the lifting here. The front-end will act accordingly given those two cases.

Reply:
That’s a useful breakdown, but it doesn’t quite cover the situation I was thinking about. I realize my original wording wasn’t super clear.

I’m thinking about the situation where we get an error from Git**b that isn’t the `403` (I think) that we’re expecting. Say we get a `5xx`, or the request times out, or there’s a DNS blip.

That might be something we’re happy to ignore for now; I’m just making sure we ignore it explicitly.